text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,187 @@
+# Meta
+[meta]: #meta
+- Name: Buildpack URIs
+- Start Date: 02/07/2020
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: "N/A"
+
+# Summary
+[summary]: #summary
+
+A set of predefined URIs for referring to buildpacks from various sources (protocols). 
+
+_Please note that the scheme denotes protocol and not type._
+
+# Motivation
+[motivation]: #motivation
+
+Given that we support a multitude of sources for buildpacks, a certain level of ambiguity has naturally presented
+itself. We currently _try_ to resolve the various sources through a complex set of rules yet the ambiguity to users
+still exists. In config files, this ambiguity is at times resolved by introducing a separate key name to signify
+the source (protocol).
+
+The aim of this RFC is to propose a set of URI schemes that could be used in both CLI arguments and config files
+while solving the ambiguity problem.
+
+A few references to further express the problem:
+
+- [A poll](https://buildpacks.slack.com/archives/C94UJCNV6/p1580920669081100) to understand what users expect when
+providing buildpacks and there is overlap of sources.
+- [Issue](https://github.com/buildpacks/pack/issues/489) to provide additional clarity in logs due to ambiguity. 
+- [A conversation](https://buildpacks.slack.com/archives/CJ6B92ZSB/p1578502454017000) when trying to determine how to
+provide reference to all buildpacks in builder.
+- [Issue](https://github.com/buildpacks/pack/issues/376) where an `image` key was added to differentiate between
+directory and image. 
+ 
+# What it is
+[what-it-is]: #what-it-is
+
+### Scheme
+
+| Name | Format | Examples |
+|--- |--- |--- |
+| Relative | `<path>` | `./my/buildpack.tgz`<br>`/home/user/my/buildpack.tgz`
+| Filesystem | `file://[<host>]/<path>` | `file:///my/buildpack.tgz`<br>`file:///home/user/my/buildpack.tgz`
+| URL | `http[s]://<host>/<path>` | `http://example.com/my/buildpack.tgz`<br>`https://example.com/my/buildpack.tgz`  
+| Docker | `docker://[<host>]/<path>[:<tag>⏐@<digest>]` | `docker://gcr.io/distroless/nodejs`<br>`docker:///ubuntu:latest`<br>`docker:///ubuntu@sha256:45b23dee08...`
+| CNB Registry | `cnb://[<host>]/[<id>[:<version>]]` |  `cnb:///my-org/my-bp`<br>`cnb://index.buildpack.io/my-org/my-bp`
+| CNB Builder Resource | `urn:cnb:builder[:<id>[:<version>]]` | `urn:cnb:builder`<br>`urn:cnb:builder:bp.id`<br>`urn:cnb:builder:bp.id:bp.version`
+| CNB Registry Resource | `urn:cnb:registry[:<id>[:<version>]]` | `urn:cnb:registry:bp.id`<br>`urn:cnb:registry:bp.id:bp.version`
```

Comment:
In pack we already use `<id>@<version>`. Should we make this `urn:cnb:registry:<id>@<version>` instead of the `:`?

Reply:
I have no preference but do like the simplicity of using colon `:` as a general segment  separator.

Reply:
agree. but i'd hate to have two different way's of handling it:

`pack build --buildpack example/foo@1.0`

and

`pack build --buildpack urn:cnb:registry:example/foo:1.0`

Reply:
Two thoughts.

1. We could migrate pack to always use colon `:`
2. `example/foo:1.0` does have the undertone of being "mutable" where `@` is "immutable" thanks to docker and node. Is there anything we want to carry forward from that?

Does that help in any way?

Reply:
From WG, let's go with `@` for URNs and CNB Registry. It implies immutability, is inline current implementation, and treats `id` + `version` as a unit vs separate segments.